app/uk/gov/hmrc/transitmovements/controllers/ArrivalsController.scala

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2022 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package uk.gov.hmrc.transitmovements.controllers
+
+import akka.stream.Materializer
+import akka.stream.scaladsl.FileIO
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import play.api.Logging
+import play.api.libs.Files.TemporaryFileCreator
+import play.api.libs.json.Json
+import play.api.mvc.Action
+import play.api.mvc.ControllerComponents
+import play.api.mvc.Result
+import uk.gov.hmrc.play.bootstrap.backend.controller.BackendController
+import uk.gov.hmrc.transitmovements.controllers.errors.ConvertError
+import uk.gov.hmrc.transitmovements.controllers.stream.StreamingParsers
+import uk.gov.hmrc.transitmovements.models.ArrivalId
+import uk.gov.hmrc.transitmovements.models.EORINumber
+import uk.gov.hmrc.transitmovements.models.MessageType
+import uk.gov.hmrc.transitmovements.models.MovementType
+import uk.gov.hmrc.transitmovements.models.formats.PresentationFormats
+import uk.gov.hmrc.transitmovements.models.responses.ArrivalDeclarationResponse
+import uk.gov.hmrc.transitmovements.repositories.MovementsRepository
+import uk.gov.hmrc.transitmovements.services.MessageFactory
+import uk.gov.hmrc.transitmovements.services.MovementFactory
+import uk.gov.hmrc.transitmovements.services.MovementsXmlParsingService
+
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class ArrivalsController @Inject() (
+  cc: ControllerComponents,
+  movementFactory: MovementFactory,
+  messageFactory: MessageFactory,
+  repo: MovementsRepository,
+  xmlParsingService: MovementsXmlParsingService,
+  val temporaryFileCreator: TemporaryFileCreator
+)(implicit
+  val materializer: Materializer
+) extends BackendController(cc)
+    with Logging
+    with StreamingParsers
+    with ConvertError
+    with TemporaryFiles
+    with PresentationFormats {
+
+  def createArrival(eori: EORINumber): Action[Source[ByteString, _]] = Action.async(streamFromMemory) {
+    implicit request =>
+      withTemporaryFile {
+        (temporaryFile, source) =>
+          (for {
+            declarationData <- xmlParsingService.extractDeclarationData(source).asPresentation
```

Comment:
Just thought about this, an arrival is going to pull out slightly different data, so this isn't going to work. This is what Sandeep was talking about in clarification.

We'll want to extract an `ArrivalData` or something like that - which extract the same things as the departure declaration, plus the MRN. Handily, [XmlParsers](https://github.com/hmrc/transit-movements/blob/main/app/uk/gov/hmrc/transitmovements/services/XmlParsers.scala#L35-L58) has extractors for each one, but only one of them has been parameterised to work based on the message type.

[You'll need to make one of these for Arrival Notifications](https://github.com/hmrc/transit-movements/blob/36d97c24044c623a9bc54c2ab779a148674282bd/app/uk/gov/hmrc/transitmovements/services/DeparturesXmlParsingService.scala#L65-L84), but it'll be a three item ZipWith that gets all the preparationDateTime, MRN and movement EORI.

Reply:
updated the logic arrival to extract the data